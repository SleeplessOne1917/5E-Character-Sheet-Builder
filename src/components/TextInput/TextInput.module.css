.input-and-error-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    width: fit-content;
}

.input-and-label-container {
    display: flex;
    position: relative;
    flex-direction: column;
    align-items: center;
    margin-top: 1rem;
}

.input-container {
    padding-top: 1.2rem;
    display: flex;
    align-items: center;
}

.input {
    font-family: inherit;
    font-size: 1.1rem;
    background-color: var(--heaven-light);
    border: 1px solid var(--davy);
    border-radius: 0.5rem;
    color: var(--davy);
    margin-top: 0.25rem;
}

.input:focus-visible,
.input:active {
    outline: var(--focus-outline);
}

.input::placeholder {
    color: transparent;
}

.label {
    position: absolute;
    bottom: 0.25rem;
    left: 0.5rem;
    color: var(--davy-light);
    cursor: text;
    pointer-events: none;
    height: min-content;
}

.label-selected {
    color: var(--davy) !important;
    bottom: 0 !important;
    left: 0 !important;
    top: 0;
    font-size: 0.9rem !important;
    pointer-events: auto !important;
}

.light-label.label-selected {
    color: var(--heaven-light) !important;
}

.error {
    color: var(--wine-red);
    width: 90%;
    display: flex;
    justify-content: center;
    padding: 0.5rem;
    font-size: 0.8rem;
}

.input-error {
    border: 3px solid var(--error-red) !important;
}

.eye {
    width: 1.3rem;
    height: 1.3rem;
    fill: transparent;
    stroke: var(--davy);
    margin-left: 0.5rem;
    cursor: pointer;
}

.eye:active, .eye:focus-visible {
    fill: var(--highlight)
}

.eye:focus-visible {
    outline: var(--focus-outline);
}


@media screen and (min-width: 768px) {
    .input {
        font-size: 1.6rem
    }

    .label {
        font-size: 1.6rem;
    }

    .label-selected {
        font-size: 1.2rem !important;
    }

    .input-container {
        padding-top: 1.5rem
    }

    .error {
        font-size: 1rem;
    }

    .eye {
        height: 1.6rem;
        width: 1.6rem;
    }
}

@media screen and (min-width: 1024px) {
    .input {
        font-size: 2rem;
    }

    .label {
        font-size: 2rem;
    }

    .input-container {
        padding-top: 2rem;
    }

    .label-selected {
        font-size: 1.6rem !important;
    }

    .error {
        font-size: 1.5rem;
    }
}

@media screen and (min-width: 1280px) {
    .eye {
        height: 2rem;
        width: 2rem;
    }
}

@media (hover: hover) and (pointer: fine) {
    .eye:hover {
        fill: var(--highlight)
    }
}