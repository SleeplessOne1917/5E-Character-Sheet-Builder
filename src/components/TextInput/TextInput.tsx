import {
	ChangeEventHandler,
	FocusEventHandler,
	KeyboardEventHandler,
	ReactNode,
	useCallback,
	useState
} from 'react';
import { EyeIcon, EyeOffIcon } from '@heroicons/react/solid';

import PasswordValidator from '../PasswordValidator/PasswordValidator';
import classes from './TextInput.module.css';

type TextInputProps = {
	touched?: boolean;
	error?: string;
	id: string;
	value: string;
	label: ReactNode;
	onChange: ChangeEventHandler<HTMLInputElement>;
	onBlur: FocusEventHandler<HTMLInputElement>;
	type?: 'text' | 'password' | 'validate-password';
};

const TextInput = ({
	touched,
	error,
	id,
	value,
	label,
	onChange,
	onBlur,
	type = 'text'
}: TextInputProps) => {
	const [showPassword, setShowPassword] = useState(false);

	const toggleShowPassword = useCallback(() => {
		setShowPassword(prevState => !prevState);
	}, [setShowPassword]);

	const toggleShowPasswordKeyDown: KeyboardEventHandler<SVGSVGElement> =
		useCallback(
			event => {
				if (event.code === 'Enter' || event.code === 'Space') {
					event.preventDefault();
					event.stopPropagation();
					toggleShowPassword();
				}
			},
			[toggleShowPassword]
		);

	return (
		<div
			className={classes['input-and-error-container']}
			data-testid="text-input"
		>
			<div className={classes['input-and-label-container']}>
				<div className={classes['input-container']}>
					<input
						className={`${classes.input}${
							touched && error ? ` ${classes['input-error']}` : ''
						}`}
						id={id}
						name={id}
						type={showPassword || type === 'text' ? 'text' : 'password'}
						value={value}
						onChange={onChange}
						onBlur={onBlur}
					/>
					{type === 'text' ? (
						''
					) : showPassword ? (
						<EyeOffIcon
							className={classes.eye}
							onClick={toggleShowPassword}
							onKeyDown={toggleShowPasswordKeyDown}
							tabIndex={0}
							aria-hidden="false"
							aria-label={`Hide ${typeof label === 'string' ? label : id}`}
						/>
					) : (
						<EyeIcon
							className={classes.eye}
							onClick={toggleShowPassword}
							onKeyDown={toggleShowPasswordKeyDown}
							tabIndex={0}
							aria-hidden="false"
							aria-label={`Show ${typeof label === 'string' ? label : id}`}
						/>
					)}
				</div>
				<label
					htmlFor={id}
					className={`${classes.label}${
						value.length > 0 ? ` ${classes['label-selected']}` : ''
					}`}
				>
					{label}
				</label>
			</div>
			{type === 'validate-password' ? (
				<PasswordValidator password={value} />
			) : (
				touched && error && <div className={classes.error}>{error}</div>
			)}
		</div>
	);
};

export default TextInput;
