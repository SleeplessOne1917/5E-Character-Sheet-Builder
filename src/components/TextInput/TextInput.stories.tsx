import { ComponentMeta, ComponentStory } from '@storybook/react';

import TextInput from './TextInput';

export default {
	title: 'Components/TextInput',
	component: TextInput,
	args: {
		label: 'Label',
		id: 'id'
	}
} as ComponentMeta<typeof TextInput>;

const Template: ComponentStory<typeof TextInput> = args => (
	<TextInput {...args} />
);

export const Default = Template.bind({});
Default.args = {
	value: ''
};

export const Password = Template.bind({});
Password.args = {
	type: 'password',
	label: 'Password',
	value: 'abcde'
};

export const ValidatePassword = Template.bind({});
ValidatePassword.args = {
	value: 'abcde',
	type: 'validate-password',
	label: 'Password'
};

export const WithError = Template.bind({});
WithError.args = {
	error: 'This is an error',
	touched: true,
	value: 'abcd1234'
};
