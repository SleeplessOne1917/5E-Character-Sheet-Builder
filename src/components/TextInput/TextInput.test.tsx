import * as stories from './TextInput.stories';

import { render, screen } from '@testing-library/react';

import TextInput from './TextInput';
import { composeStories } from '@storybook/react';
import userEvent from '@testing-library/user-event';

const renderTestCases = Object.values(composeStories(stories));

it.each(renderTestCases)('renders correctly', Story => {
	render(<Story />);

	expect(screen.getByTestId('text-input')).toMatchSnapshot();
});

it('calls onChange when expected', async () => {
	const onChangeMock = jest.fn();
	render(
		<TextInput
			id="id"
			label="Label"
			onBlur={() => {}}
			onChange={onChangeMock}
			value=""
		/>
	);

	await userEvent.type(screen.getByLabelText(/Label/i), 'abcde');

	expect(onChangeMock).toHaveBeenCalled();
});

it('calls onBlur when expected', async () => {
	const onBlurMock = jest.fn();
	render(
		<TextInput
			id="id"
			label="Label"
			onBlur={onBlurMock}
			onChange={() => {}}
			value=""
		/>
	);

	await userEvent.type(screen.getByLabelText(/Label/i), 'abcde');
	await userEvent.tab();

	expect(onBlurMock).toHaveBeenCalled();
});
