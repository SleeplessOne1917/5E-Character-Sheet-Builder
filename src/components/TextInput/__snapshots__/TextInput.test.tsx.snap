// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="input-and-error-container"
  data-testid="text-input"
>
  <div
    class="input-and-label-container"
  >
    <div
      class="input-container"
    >
      <input
        class="input"
        id="id"
        name="id"
        type="text"
        value=""
      />
    </div>
    <label
      class="label"
      for="id"
    >
      Label
    </label>
  </div>
</div>
`;

exports[`renders correctly 2`] = `
<div
  class="input-and-error-container"
  data-testid="text-input"
>
  <div
    class="input-and-label-container"
  >
    <div
      class="input-container"
    >
      <input
        class="input"
        id="id"
        name="id"
        type="password"
        value="abcde"
      />
      <svg
        aria-hidden="false"
        aria-label="Show Password"
        class="eye"
        fill="currentColor"
        tabindex="0"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 12a2 2 0 100-4 2 2 0 000 4z"
        />
        <path
          clip-rule="evenodd"
          d="M.458 10C1.732 5.943 5.522 3 10 3s8.268 2.943 9.542 7c-1.274 4.057-5.064 7-9.542 7S1.732 14.057.458 10zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <label
      class="label label-selected"
      for="id"
    >
      Password
    </label>
  </div>
</div>
`;

exports[`renders correctly 3`] = `
<div
  class="input-and-error-container"
  data-testid="text-input"
>
  <div
    class="input-and-label-container"
  >
    <div
      class="input-container"
    >
      <input
        class="input"
        id="id"
        name="id"
        type="password"
        value="abcde"
      />
      <svg
        aria-hidden="false"
        aria-label="Show Password"
        class="eye"
        fill="currentColor"
        tabindex="0"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 12a2 2 0 100-4 2 2 0 000 4z"
        />
        <path
          clip-rule="evenodd"
          d="M.458 10C1.732 5.943 5.522 3 10 3s8.268 2.943 9.542 7c-1.274 4.057-5.064 7-9.542 7S1.732 14.057.458 10zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <label
      class="label label-selected"
      for="id"
    >
      Password
    </label>
  </div>
  <div
    class="container"
    data-testid="container"
  >
    <div
      class="validator no-match"
    >
      <svg
        aria-hidden="true"
        class="icon x"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
       
      At least 8 characters long
    </div>
    <div
      class="validator match"
    >
      <svg
        aria-hidden="true"
        class="icon check"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
          fill-rule="evenodd"
        />
      </svg>
       
      At least 1 lowercase letter
    </div>
    <div
      class="validator no-match"
    >
      <svg
        aria-hidden="true"
        class="icon x"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
       
      At least 1 uppercase letter
    </div>
    <div
      class="validator no-match"
    >
      <svg
        aria-hidden="true"
        class="icon x"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
       
      At least 1 number
    </div>
    <div
      class="validator no-match"
    >
      <svg
        aria-hidden="true"
        class="icon x"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
       
      At least 1 special character
    </div>
  </div>
</div>
`;

exports[`renders correctly 4`] = `
<div
  class="input-and-error-container"
  data-testid="text-input"
>
  <div
    class="input-and-label-container"
  >
    <div
      class="input-container"
    >
      <input
        class="input input-error"
        id="id"
        name="id"
        type="text"
        value="abcd1234"
      />
    </div>
    <label
      class="label label-selected"
      for="id"
    >
      Label
    </label>
  </div>
  <div
    class="error"
  >
    This is an error
  </div>
</div>
`;
